人工智能/人工智能数学基础/人智数学基础实验/2026春实验三/实验三代码/newton.py

```
@@ -0,0 +1,213 @@
+        step = torch.linalg.solve(hess_reg, grads)  # 解 H * step = g
+    except:
+        step = torch.linalg.lstsq(hess_reg, grads).solution
+    x_new = x - lr * step[0].item()
+    y_new = y - lr * step[1].item()
+    return x_new, y_new
+
+# 三次正则化牛顿法（求解 min_p f+g^T p + 1/2 p^T H p + sigma/3 ||p||^3）
+def cubic_newton_step(x, y, func, sigma=1.0, **kwargs):
+    """
+    使用 pytorch-minimize 库求解三次正则化子问题。
+    该方法利用信赖域精确算法 (trust-exact)，能有效处理非凸问题。
```

Ruff (0.15.17)

[warning] Docstring contains ambiguous `，` (FULLWIDTH COMMA). Did you mean `,` (COMMA)?

(RUF002)